beps/0010-scaffolder-templating-in-parameters/README.md

```diff
@@ -0,0 +1,212 @@
+-->
+
+This BEP proposes to add support for templating syntax in the `parameters` schema of a scaffolder template.
+This will allow users to define properties in the JSON Schema which are templated from current values that have been collected from the user already.
+This can be useful when you want to use a value that has already been collected as a default value in another field.
+
+For example:
+
+```yaml
+apiVersion: scaffolder.backstage.io/v1beta3
+kind: Template
+metadata:
+  name: my-template
+spec:
+  parameters:
+    - title: Some input
+      description: Get some info from the user
+      properties:
+        name:
+          type: string
+          default: Test
+        description:
+          type: string
+          default: ${{ parameters.name or "unknown" }}-description
+```
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this BEP. Describe why the change is important and the benefits to users.
+-->
+
+Inclusive of the initial RFC there's been a swarm of issues that are requesting this feature, and we want to align on the implementation and design of this feature.
+
+See the following:
+
+- https://github.com/backstage/backstage/issues/16275
+- https://github.com/backstage/backstage/pull/23283
+- https://github.com/backstage/backstage/issues/19597
+- https://github.com/backstage/backstage/issues/20533
+- https://github.com/backstage/backstage/pull/17746
+
+There's some ideas for introducing a templating syntax for both templating into the `parameters` schema, and also being able to pass through some templating strings to underlying field extensions that can use those templating strings.
+We want to align here so that we're not going to have those conflict or compete, and create a standard for how to achieve templating in both circumstances.
+
+### Goals
+
+<!--
+List the specific goals of the BEP. What is it trying to achieve? How will we
+know that this has succeeded?
+-->
+
+- This BEP will settle the implementation for the templating of fields into the JSON Schema in the `parameters` section in the scaffolder templates.
+- This BEP will settle how to pass through templating strings to underlying field extensions in a non-conflicting way.
+
+### Non-Goals
+
+<!--
+What is out of scope for this BEP? Listing non-goals helps to focus discussion
+and make progress.
+-->
+
+## Proposal
+
+The proposal is to be able to decorate the template schema server side with a context and use that to drive the form rendering client side.
+
+We can extend the `/parameter-schema` endpoint to accept a `formData` context query parameter which will be a JSON object of the current `formData` state.
+
+```diff
+export interface ScaffolderApi {
+  getTemplateParameterSchema(
+    templateRef: string,
++    formData?: string,
+  ): Promise<TemplateParameterSchema>;
+}
+```
+
+```diff
+ router
+    .get(
```

Comment:
might want to support POST here if the formData can become large (eg when having file inputs or entity pickers that match a large part of the catalog or whatnot)